packages/core/types/index.d.ts

```diff
@@ -77,25 +77,30 @@ export interface ClassNamesProps<Theme> {
  * @desc
  * JSX generic are supported only after TS@2.9
  */
-export function ClassNames<Theme extends {} = any>(
+export function ClassNames<Theme extends {} = Emotion.Theme>(
   props: ClassNamesProps<Theme>
 ): ReactElement
 
 declare module 'react' {
   interface DOMAttributes<T> {
-    css?: InterpolationWithTheme<any>
+    css?: InterpolationWithTheme<Emotion.Theme>
   }
 }
 
 declare global {
+  namespace Emotion {
+    // tslint:disable-next-line: no-empty-interface strict-export-declare-modifiers
+    export interface Theme {}
```

Comment:
@JakeGinnivan any reason not to go with `styled-components` approach\*?

- Where we'd `export interface DefaultTheme {}` in the module type scope
- use as default for other packages via `import { DefaultTheme } from '@emotion/core'`
- and the user could extend via:
  ```typescript
  // emotion-theme.d.ts
  import '@emotion/core'
  
  declare module '@emotion/core' {
     export interface DefaultTheme {
      your: 'theme',
      comes: 'here'
    }
  }
  ```
I feel this one as the more cleaner approach. The one thing I do dislike, is that it's tied to `@emotion/core` namespace, which actually, has nothing to do about theming.

\* The question is not limited to Jake, just that his recommendation on the issue posting was to go this route.

Reply:
Only reason I suggested what I did is it's the one I am familiar with. Express uses this to extend the Request/Response objects.

I can't say I have enough of an understanding of the styled components approach to sum up trade offs right now.

Reply:
Something which springs to mind, is the global approach may not work with babel typescript compilation?

Reply:
@JakeGinnivan hmm, I cannot seem to find any reason why it wouldn't work with babel. In either way it would need to load an additional type declaration from local codebase. 

As long as babel supports declaration merging and namespaces - _which I think it does, otherwise it's not in compliance with TS and would beat the purpose_ - either of solutions should work.

My question to you (@JakeGinnivan) was more on performance implications (though, I don't see much difference, but I'm not a TS pro by any means) and for everyone about the overall DX we expose.

Reply:
My 2c on this is that we should do this one because it doesn't matter much and thinking in modules feels better.

```tsx
declare module '@emotion/core' {
  export interface DefaultTheme {
    your: 'theme',
    comes: 'here'
  }
}
```

> The one thing I do dislike, is that it's tied to @emotion/core namespace, which actually, has nothing to do about theming.

This isn't really true, the ThemeContext lives in `@emotion/core` and you can even use that without ever using `emotion-theming`

Reply:
@mitchellhamilton that's true that `ThemeContext` can be used directly, but is it even a good thing? I understand that wrapping "enforces" you to learn some APIs - but they relate to regular APIs so much, that it's hardly noticeable. And also wrapping (by wrapping I mean `useTheme` and `ThemeProvider` helpers) gives some better encapsulation - it defines a clear way of using a particular thing and helps us with providing any extra functionality, babel transforms or error/warning messages in those - while this becomes harder to do when exporting `ThemeContext` directly.

Having those things inside a single package would make it easier to install those things and if somebody doesn't want to use those exports tree-shaking should just do its job at pruning those.